SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1

```diff
@@ -90,46 +152,72 @@ function Set-Secret
-        $sb.Append("Username: ").AppendLine($Secret.UserName)
-        $sb.Append("login: ").AppendLine($Secret.UserName)
+    
+    
+    if ($Secret -is [string]) {
+        $Secret = @{URL = ' http://sn';Notes = $Secret}
+    } elseif ($Secret -is [pscredential]) {
+        $Secret = @{Username = $Secret.Username; Password = $Secret.GetNetworkCredential().password}
     }
 
-    if($Secret.Password) {
-        $pass = $Secret.Password
-        if ($Secret -is [pscredential]) {
-            $pass = $Secret.GetNetworkCredential().password
-        } elseif ($pass -is [securestring]) {
-            $pass = $pass | ConvertFrom-SecureString
-        }
 
-        $sb.Append("Password: ").AppendLine($pass)
-        $sb.Append("password: ").AppendLine($pass)
-        $sb.Append("sudo_password: ").AppendLine($pass)
-    }
+    
+    if ($Secret -is [hashtable]){
+        $SpecialKeys = @('Language', 'NoteType', 'Notes')
 
-    if($Secret.URL) {
-        $sb.Append("URL: ").AppendLine($Secret.URL)
-    }
+        if ($Secret.Keys.count -eq 1 -and $null -ne $Secret.Notes) {
+            $Secret.URL = 'http://sn'
+        }
 
-    if($Secret.Notes) {
-        $sb.AppendLine("Notes:").AppendLine($Secret.Notes)
-    }
+        $Keys = $Secret.Keys.Where({$_ -notin $SpecialKeys })
+        foreach ($k in $Keys) {
+            if ($Secret.$k -is [securestring]) {
+                $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.$k)
+                $Secret.$k = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+            }
+            [Void]($sb.AppendLine("$($k): $($Secret.$k)"))
+        }
 
+        # Notes need to be on a new line
+        if ($null -ne $Secret.Notes) {
+            if ($Secret.Notes -is [securestring]) {
+                $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.Notes)
+                $Secret.Notes = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+            }
+            [Void]($sb.AppendLine("Notes: `n$($Secret.Notes)"))
+        }
+    } 
+    
     try {
-        $sb.ToString() | Invoke-lpass 'add', $Name, '--non-interactive'
-    } catch {
+        $res = Invoke-lpass 'show', '--sync=now', '--name', $Name, '2>/dev/null' -ErrorAction Stop 
+        $SecretExists = $null -ne $res 
+
+        if ($SecretExists) {
+            Write-Verbose "Editing secret" 
+            $sb.ToString() | Invoke-lpass 'edit', '--non-interactive', $Name
+        } else {
+            Write-Verbose "Adding new secret" 
+            $NoteTypeArgs = @()
+            if ($null -ne $Secret.NoteType) {
+                if ($Secret.NoteType -is [securestring]) { 
+                    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.NoteType)
+                    $Secret.NoteType = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+                }
+                $NoteTypeArgs = @("--note-type=$($Secret.NoteType)")
```

Comment:
```suggestion
                $NoteTypeArgs += "--note-type=$($Secret.NoteType)"
```
If you're allocating a new array anyway, might as well do it this way.

Reply:
It's a little bit cleaner but it feels like I'm encouraging anybody that would want to expand this array later on for any reason to reuse `+=`. 

That's why I never used `+=` for array, even once.

Reply:
Done anyway :)